### DISCUSSION TITLE
[week22] useState 關於非同步的觀念

### DISCUSSION THREAD

**Discussion by chengcheng1231** [Lidemy/forum#254] [前端相關問題] [answered]
最近在回頭複習 useState ，但發現有個觀念過去只是知道有這樣的情況，但沒有實際思考到底為什麼這樣改會得到我們要的效果，所以想提問：

useState 的 state update function 是「非同步」執行的，像以下情況，短時間內連續按下按鈕會不是預期的結果，執行 `setValue(value + 1)` 並不會依照我們預期印出 2、3、4...。
```JavaScript
export default function App() {
  const [ value, setValue] = useState(1)

  function handleClick() {
    setTimeout(()=>{setValue(value + 1)},3000)

  }
  console.log(value)

  return (
    <div className="App">
      <button onClick={handleClick}>按鈕</button>
    </div>
  );
}
```
連續快速點擊按鈕三次，但印出來的只有兩次：
<img width="287" alt="pic1" src="https://user-images.githubusercontent.com/36998811/158047568-58a1d790-eb85-4a21-8166-9e9a51667f77.png">

若中間更新的部分用 func 改寫成以下就能改善，變成「同步」的效果，我去保證了執行順序，它看起來就是update 完 state 後再做下一件事

```JavaScript
function handleClick() {
    setTimeout(()=>{setValue(value => value + 1)},3000)
  }
```

一樣是連續快速點擊按鈕三次（甚至更多次）：
<img width="323" alt="pic5" src="https://user-images.githubusercontent.com/36998811/158048037-e2d5c282-3e66-4f02-a02e-a4de6b2328fe.png">

附上：[codesandbox](https://codesandbox.io/s/my-react-pud39p?file=/src/App.js)

我覺得了解到這樣其實已經能解決我工作上的問題，但這兩種情況都有讓我百思不得其解的地方~

1. 在未改寫之前我連續快速點擊按鈕三次 （甚至更多次），但 console 出來都會是兩次一樣的值，我了解為什麼會是一樣的值，但為什麼是兩次？而不是等於點擊的次數？，是否是因為非同步的原因，無論點擊多少次，setCount 帶上的參數 value 會是 1， 也就是 setCount(2)，所以都會被看做是一樣的，因此只會 update 一次，但如果是這樣，應該只會 console 一次，但結果偏偏就是兩次  @@

2. 改寫之後為什麼把 func 傳入 setValue 就能變成同步的效果，是否要去看 useState 它做 update 的 func 原本是怎麼處理的？

以上不知道是我哪邊的基礎觀念出了差錯，還是我不用知道這麼多，希望老師還會來看這邊的討論區 ><

**Comment by aszx87410** [comment, marked as answer]
1. 這個我倒是不知道，也沒有思考過，不太確定為什麼會是兩次，因為照理來說確實是一次比較合理
2. 對，因為這跟他底層 update 的機制有關

如果想知道的話，中國有不少文章會講這類型比較底層的運作，可以用關鍵字 `react 異步 更新` 之類的去找，應該可以找到不少資源
